src/librustc_mir/interpret/eval_context.rs

```diff
@@ -547,60 +551,122 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         }
     }
 
-    pub(super) fn pop_stack_frame(&mut self) -> InterpResult<'tcx> {
-        info!("LEAVING({}) {}", self.cur_frame(), self.frame().instance);
+    /// Pops the current frame from the stack, deallocating the
+    /// memory for allocated locals.
+    ///
+    /// If `unwinding` is `false`, then we are performing a normal return
+    /// from a function. In this case, we jump back into the frame of the caller,
+    /// and continue execution as normal.
+    ///
+    /// If `unwinding` is `true`, then we are in the middle of a panic,
+    /// and need to unwind this frame. In this case, we jump to the
+    /// `cleanup` block for the function, which is responsible for running
+    /// `Drop` impls for any locals that have been initialized at this point.
+    /// The cleanup block ends with a special `Resume` terminator, which will
+    /// cause us to continue unwinding where we left off.
+    pub(super) fn pop_stack_frame(
+        &mut self,
+        unwinding: bool
+    ) -> InterpResult<'tcx> {
+        info!("LEAVING({}) {} (unwinding = {})",
+            self.cur_frame(), self.frame().instance, unwinding);
+
+        // Sanity check `unwinding`.
+        assert_eq!(
+            unwinding,
+            match self.frame().block {
+                None => true,
+                Some(block) => self.body().basic_blocks()[block].is_cleanup
+            }
+        );
+
         ::log_settings::settings().indentation -= 1;
         let frame = self.stack.pop().expect(
             "tried to pop a stack frame, but there were none",
         );
-        M::stack_pop(self, frame.extra)?;
-        // Abort early if we do not want to clean up: We also avoid validation in that case,
+        let stack_pop_info = M::stack_pop(self, frame.extra)?;
+        match (unwinding, stack_pop_info) {
+            (true, StackPopInfo::StartUnwinding) =>
+                bug!("Attempted to start unwinding while already unwinding!"),
+            (false, StackPopInfo::StopUnwinding) =>
+                bug!("Attempted to stop unwinding while there is no unwinding!"),
+            _ => {}
+        }
+
+        // Usually we want to clean up (deallocate locals), but in a few rare cases we don't.
+        // In that case, we return early. We also avoid validation in that case,
         // because this is CTFE and the final value will be thoroughly validated anyway.
-        match frame.return_to_block {
-            StackPopCleanup::Goto(_) => {},
-            StackPopCleanup::None { cleanup } => {
-                if !cleanup {
-                    assert!(self.stack.is_empty(), "only the topmost frame should ever be leaked");
-                    // Leak the locals, skip validation.
-                    return Ok(());
-                }
+        let cleanup = unwinding || match frame.return_to_block {
+            StackPopCleanup::Goto{ .. } => true,
+            StackPopCleanup::None { cleanup, .. } => {
+                cleanup
```

Comment:
I preferred the previous code's scheme where the `None` arm had the cleanup check. Now it could just be `!(unwinding || cleanup)`. The way the code is currently written it took me a minute to figure out the logic and realize that `Goto` never goes into the cleanup, no matter the value of `unwinding`

Reply:
Hm, I preferred giving the boolean a name.^^

We could however factor the `unwinding ||` into the `match`:
```rust
let cleanup = match frame.return_to_block {
            StackPopCleanup::Goto{ .. } => true,
            StackPopCleanup::None { cleanup, .. } => {
                bug!("Encountered StackPopCleanup::None while unwinding");
                cleanup
            }
}
```

> Goto never goes into the cleanup,

You mean "always", right?

Reply:
@Aaron1011 could you refactor this as @oli-obk suggested?

Here's a way for us to avoid the double-matching (I think this is closer to what @oli-obk suggested):
```rust
let (cleanup, next_block) = match frame.return_to_block {
    StackPopCleanup::Goto{ ret, unwind } => (true, Some(if unwinding { unwind } else { ret })),
    StackPopCleanup::None { cleanup, .. } => (cleanup, None)
};
```
Then we do not need to match on `return_to_block` every again: in the `if cur_unwinding {` branch you can `next_block.expect("Encountered StackPopCleanup::None while unwinding")`, and in the `else` branch you do `if let Some(next_block) = next_block { ...`.

Reply:
Hm, actually this still doesn't properly encode that if we do not cleanup we also do not have a next block... but that's probably not worth putting in the type, I'd just `assert!(next_block.is_none())` in the `if !cleanup {` branch.